### ISSUE TITLE
UncommentedMainCheck: instance void main() not flagged in compact source files (JEP 512)

### ISSUE THREAD

**Issue by vivek-0509** [checkstyle/checkstyle#20220]
Parent Issue #19971

I have read check documentation: https://checkstyle.org/checks/uncommentedmain.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > config.xml <<'EOF'
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="UncommentedMain"/>
  </module>
</module>
EOF

vivek@Viveks-MacBook-Air checkstyle % cat > InputUncommentedMainCompact.java <<'EOF'
void main() {
    System.out.println("hello");
}
EOF

vivek@Viveks-MacBook-Air checkstyle % RUN_LOCALE="-Duser.language=en -Duser.country=US"
java $RUN_LOCALE -jar target/checkstyle-13.5.0-SNAPSHOT-all.jar -c config.xml InputUncommentedMainCompact.java

Starting audit...
Audit done.
```

Same code inside an ordinary class (using the traditional `public static void main(String[] args)` form):

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > OrdinaryUncommentedMain.java <<'EOF'
public class OrdinaryUncommentedMain {
    public static void main(String[] args) {
        System.out.println("hello");
    }
}
EOF

vivek@Viveks-MacBook-Air checkstyle % java $RUN_LOCALE -jar target/checkstyle-13.5.0-SNAPSHOT-all.jar -c config.xml OrdinaryUncommentedMain.java

Starting audit...
[ERROR] OrdinaryUncommentedMain.java:2:5: Uncommented main method found. [UncommentedMain]
Audit done.
Checkstyle ends with 1 errors.
```

---

**Describe what you expect in detail.**

UncommentedMainCheck must recognise an instance `void main()` (no `public`, no `static`, no parameters) as a valid program entry point in a JEP 512 compact source file. The check's `checkModifiers` currently requires both `LITERAL_PUBLIC` and `LITERAL_STATIC` to consider a method as a main method. JEP 512 makes instance `void main()` a valid entry point, so the check must accept this form.

Expected on the compact input: 1 violation `Uncommented main method found.` on `main`, identical to the ordinary class above.

**Comment by Carbon14-48** [comment]
on it

**Comment by romani** [comment]
@Carbon14-48 , 

> UncommentedMainCheck 
Rationale: A main method is often used for debugging purposes. When debugging is finished, developers often forget to remove the method, which changes the API and increases the size of the resulting class or JAR file. Except for the real program entry points, all main methods should be removed or commented out of the sources. 

Whole reason of compact source is to have `main` method as entry point. All applications have main method, all of them, but by this Check they control that only one class is responsible for entry point to application, no back door.
in compact source there is no other classes and files, we have MyScript.java and it must have entry point method - main. There is no other reason in such compact source.

from Goggle AI:
>Single-File Isolation: The primary goal of compact source files is to streamline prototyping, scripting, and learning by eliminating boilerplate code. They are explicitly designed to act as independent, single-file source-code programs.


from  https://openjdk.org/jeps/512 :
> Goals
    Offer a smooth on-ramp to Java programming, so that instructors can introduce concepts in a gradual manner.
    Help students write simple programs in a concise manner, and grow their code gracefully as their skills grow.
    Reduce the ceremony of writing other kinds of small programs, such as scripts and command-line utilities.
    Do not introduce a separate dialect of the Java language.
    Do not introduce a separate toolchain. Small Java programs should be compiled and run with the same tools as large programs.

With all this in mind Check must skip compact sources by design, and I am completely ok to reference this in Check documentation.

**Comment by Carbon14-48** [comment]
@romani Thanks for the detailed explanation it  makes perfect sense. I completely missed the broader context of why the check exists in the first place!

**Comment by romani** [comment]
doc update is merged.
